ext/mysql2/result.c

```diff
@@ -498,10 +526,12 @@ static VALUE rb_mysql_result_fetch_row(VALUE self, MYSQL_FIELD * fields, const r
     wrapper->numberOfFields = mysql_num_fields(wrapper->result);
     wrapper->fields = rb_ary_new2(wrapper->numberOfFields);
   }
-  if (args->asArray) {
-    rowVal = rb_ary_new2(wrapper->numberOfFields);
-  } else {
+  
+  if (args->rowsAs == AS_HASH) {
     rowVal = rb_hash_new();
+  } else /* array or struct */ {
+    /* struct uses array as an intermediary */
+    rowVal = rb_ary_new2(wrapper->numberOfFields);
```

Comment:
These don't get GC marked, why does the struct variant need to be GC marked? (`w->rowStruct` above) (Or conversely, should these also get GC marked?)

Reply:
I admit I'm hazy on when something needs to be gc-marked and when it doesn't. The "prototype" struct needs to be marked so that it'll stick around while control hops between the gem and ruby, while enumerating the results. Each row result struct, tho', is an instance of this prototype struct and is made right here, just like the hash and array, and returned to the client asap and will be gc marked there. That's my understanding anyway.